opencv/c_extension/fast_image_processing.cpp

```diff
@@ -0,0 +1,229 @@
+#include <string>
+#include <iostream>
+#include <vector>
+#include <math.h> // atan
+#include <cmath> // std::abs
+
+#include <pybind11/pybind11.h>
+
+#include <Eigen/Core>
+#include <Eigen/Dense>
+
+#include <opencv2/core/core.hpp>
+#include <opencv2/core/eigen.hpp>
+#include <opencv2/highgui/highgui.hpp>
+#include <opencv2/imgproc.hpp>
+
+#define CONVERT_TO_INT true
+#define MAX_WIDTH 320
+#define PI 3.14159265
+#define REF_ANGLE float(- PI / 2)
+#define MAX_ANGLE float(PI / 4)
+#define IM_WIDTH MAX_WIDTH
+#define N_REGIONS 3
+
+#include "ndarray_converter.h"
+
+using namespace Eigen;
+using EigenMat = Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>;
+
+namespace py = pybind11;
+
+bool initialized = false;
+
+
+cv::Mat centroids;
+cv::Mat resized = cv::Mat::ones(20, 80, CV_32FC3);
+
+cv::Mat w1 = cv::Mat::ones(80*20*3, 20, CV_32F).t();
+cv::Mat b1 = cv::Mat::ones(20, 1, CV_32F).t();
+
+cv::Mat w2 = cv::Mat::ones(20, 4, CV_32F).t();
+cv::Mat b2 = cv::Mat::ones(4, 1, CV_32F).t();
+
+cv::Mat w3 = cv::Mat::ones(4, 1, CV_32F).t();
+cv::Mat b3 = cv::Mat::ones(1, 1, CV_32F).t();
+
+
+// Regions of interest (ROIs)
+cv::Rect R0(0, 150, MAX_WIDTH, 50);
+cv::Rect R1(0, 125, MAX_WIDTH, 50);
+cv::Rect R2(0, 100, MAX_WIDTH, 50);
+cv::Rect R3(0, 75, MAX_WIDTH, 50);
+cv::Rect R4(0, 50, MAX_WIDTH, 50);
+// ROI used
+std::vector<cv::Rect> REGIONS({R1, R2, R3});
+
+/**
+ * Set the weights of the neural network
+ * it should be called from python
+ */
+void setWeights(cv::Mat w1_, cv::Mat b1_, cv::Mat w2_, cv::Mat b2_, cv::Mat w3_, cv::Mat b3_)
+{
+  w1 = w1_.clone().t();
```

Comment:
why clone, and why transpose?

Reply:
Whithout cloning i had seg fault, transpose is more for convenience for later matrix multiplication